### ISSUE TITLE
AQL WITH with bind parameters results in "access after data-modification" for two independent UPSERTs

### ISSUE THREAD

**Issue by awlayton** [arangodb/arangodb#4677]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [ ] 3.1
- [ ] 3.2
- [x] 3.3
- [ ] self-compiled devel branch

Mode:
- [x] Cluster
- [ ] Single-Server

Storage-Engine:
- [ ] mmfiles
- [x] rocksdb

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [x] Linux 
  - [ ] Debian .deb
  - [ ] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [x] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

### this is an AQL-related issue:
[ ] I'm using graph features

I'm issuing AQL via:
- [x] web interface with this browser: Vivaldi 1.14.1077.50     running on this OS: Ubuntu 16.04.4
- [ ] arangosh
- [ ] this Driver:

I've run `db._explain("<my aql query>")` and it didn't shed more light on this.
The AQL query in question is:

The issue can be reproduced using this dataset:

Please provide a way to create the dataset to run the above query on; either by a gist with an arangodump, or `db.collection.save({my: "values"}) statements. If it can be reproduced with one of the ArangoDB example datasets, it's a plus.

When I try the below query (either execute or explain), I get the error:
`Explain: access after data-modification by collection 'graphNodes' (while loading collections)`
```js
WITH @@resources, @@graphNodes
LET res = @obj                                                   
LET foo = (
    LET nodekey = CONCAT('resources:', res._key)                             
    UPSERT { '_key': nodekey }                                               
    INSERT {                                                                 
    '_key': nodekey,                                                       
    'is_resource': true,                                                   
    'resource_id': res._id                                                 
    }                                                                        
    UPDATE {}                                                                
    IN @@graphNodes
)
UPSERT { '_key': res._key }                                              
INSERT res                                                               
UPDATE res                                                               
IN @@resources
RETURN OLD._oada_rev 
```

However, if I don't use the bind parameters for the WITH,
the below query runs and works as expected.
```js
WITH resources, graphNodes
LET res = @obj                                                   
LET foo = (
    LET nodekey = CONCAT('resources:', res._key)                             
    UPSERT { '_key': nodekey }                                               
    INSERT {                                                                 
    '_key': nodekey,                                                       
    'is_resource': true,                                                   
    'resource_id': res._id                                                 
    }                                                                        
    UPDATE {}                                                                
    IN @@graphNodes
)
UPSERT { '_key': res._key }                                              
INSERT res                                                               
UPDATE res                                                               
IN @@resources
RETURN OLD._oada_rev 
```

Is this a bug or is WITH not supposed to be used with bind parameters? I couldn't find anything definitive either way in the docs.

Thanks

**Comment by jsteemann** [comment]
Should be fixed by the above PRs.
The one for 3.3.4 can be found here: https://github.com/arangodb/arangodb/pull/4678

**Comment by OmarAyo** [comment]
Hi @awlayton ,

Thanks for reporting this bug

We have released 3.3.4, and it includes a fix for it
I am closing this ticket now. Please feel free to reopen if the problem persists
